Multiple memory handling fixes

michalpasztamobica opened the pull request:
### Description

Based on valgrind reports running on unit tests a couple of changes were introduced.
Some unit tests and stubs were not deleting allocated objects.
Production code:
* TLSSocketWrapper frees allocated cert buffer in case of error from mbedtls,
* nsapi_addr has a mem_init() function, initializing all of its memory.

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@SeppoTakalo 
@JuhPuur 
@VeijoPesonen 
@mtomczykmobica 
@tymoteuszblochmobica

SeppoTakalo commented on features/netsocket/nsapi_types.h:
```diff
@@ -190,6 +190,37 @@ typedef struct nsapi_addr {
     uint8_t bytes[NSAPI_IP_BYTES];
+
+    /** Default constructor */
+    nsapi_addr() :
```
If you recompile, you might notice that `nsapi_types.h` is also included in C files. So this needs to be pure C99.

michalpasztamobica replied:
Thanks for pointing this out, @SeppoTakalo .
Any idea how to make sure a C structure will always be initialized? I was trying to avoid initializing every instance of nsapi_addr we have in our code...

VeijoPesonen reviewed (commented):
In the future test code, implementation and cosmetic changes could be split into separate commits for readability.

michalpasztamobica committed:
Multiple memory handling fixes

Correct memory clean-ups in multiple tests and a stub.

michalpasztamobica committed:
Multiple memory handling fixes

Based on valgrind reports running on unit tests following changes were introduced:
* TLSSocketWrapper frees allocated cert buffer in case of errors from mbedtls,
* nsapi_addr has a mem_init() function, initializing all of its memory during construction.

VeijoPesonen commented on features/netsocket/TLSSocketWrapper.cpp:
```diff
@@ -130,12 +132,16 @@ nsapi_error_t TLSSocketWrapper::set_client_cert_key(const void *client_cert, siz
     if ((ret = mbedtls_x509_crt_parse(crt, static_cast<const unsigned char *>(client_cert),
                                       client_cert_len)) != 0) {
         print_mbedtls_error("mbedtls_x509_crt_parse", ret);
+        mbedtls_x509_crt_free(crt);
```
without knowing what mbedtls_x509_crt_free() does the name implies, for me, that it would delete the crt... But that just me. Take this comment just as me thinking out loud

michalpasztamobica replied:
In  a way I agree with you, but if you look close the logic of our code was:
1) crt = new;
2) mbedtls_x509_crt_init
3) mbedtls_x509_crt_free
so where is the delete corresponding to what we allocated in step 1? :)

The implementation of [mbedtls_x509_crt_free](https://github.com/ARMmbed/mbed-os/blob/3aeff7ae2578598c112d55fc127fd33af1e46f0d/features/mbedtls/src/x509_crt.c#L2644) doesn't make things easier, but in the end I think it deallocates everything aside from the poitner passed to it itself...:
```
    do
    {
        cert_prv = cert_cur;
        cert_cur = cert_cur->next;

        mbedtls_platform_zeroize( cert_prv, sizeof( mbedtls_x509_crt ) );
        if( cert_prv != crt )
            mbedtls_free( cert_prv );
    }
```

I ran a greentea TLS testsuite and it passed fine throwing no double free exceptions, so I guess we're doing the right thing in the end.
But thanks for staying alert ;)

SeppoTakalo replied:
It frees all the storage/data that Mbed TLS have allocated for it.
But it does not free the storage of the structure itself, which can be statically allocated.

Therefore we need to call `delete` on the storage.

SeppoTakalo replied:
Basically this is clear, if you look the API pair.

```
/**
 * \brief          Initialize a certificate (chain)
 *
 * \param crt      Certificate chain to initialize
 */
void mbedtls_x509_crt_init( mbedtls_x509_crt *crt );

/**
 * \brief          Unallocate all certificate data
 *
 * \param crt      Certificate chain to free
 */
void mbedtls_x509_crt_free( mbedtls_x509_crt *crt );
```

Both take pointer to the storage.
So the TLS cannot (should not) free something it has not allocated. This is common API design.
Ownership of the pointer does not transfer unless clearly specified.

SeppoTakalo commented on features/netsocket/SocketAddress.h:
```diff
@@ -91,6 +91,9 @@ class SocketAddress {
     ~SocketAddress();
 
+    /** Initialize memory */
+    void mem_init(void);
```
Move this to private section. This is not a public API.

SeppoTakalo reviewed (approved):
Looks Good.
